File: frontend/src/types/product.ts
export interface PharmacyPrice {
  pharmacy_source: string;
  site_code?: string;
  product_name?: string;
  current_price?: number;
  original_price?: number;
  discount_percentage?: number;
  discount_amount?: number;
  bank_discount_price?: number;
  bank_discount_bank_name?: string;
  bank_payment_offers?: string;
  requires_prescription?: boolean;
  product_url?: string;
  image_url?: string;
  scraped_at?: string;
}

export interface ComparisonData {
  barcode: string;
  product_name?: string;
  brand?: string;
  main_category?: string;
  image_url?: string;
  best_price?: number;
  highest_price?: number;
  savings?: number;
  pharmacies: PharmacyPrice[];
}

export interface SearchResultItem {
  group_key?: string;
  barcode?: string;
  product_name?: string;
  brand?: string;
  image_url?: string;
  main_category?: string;
  requires_prescription?: boolean;
  best_price?: number;
  pharmacy_count?: number;
}

export interface SearchResponse {
  results: SearchResultItem[];
  total: number;
  page: number;
  limit: number;
}

export interface CategoryItem {
  main_category: string;
  product_count: number;
  pharmacy_count?: number;
  min_price?: number;
  max_price?: number;
}

export interface CategoryProductItem {
  id: string;
  pharmacy_source: string;
  site_code?: string;
  barcode?: string;
  product_name?: string;
  brand?: string;
  image_url?: string;
  main_category?: string;
  current_price?: number;
  original_price?: number;
  discount_percentage?: number;
  discount_amount?: number;
  bank_discount_price?: number;
  bank_discount_bank_name?: string;
  requires_prescription?: boolean;
  product_url?: string;
  scraped_at?: string;
}

export interface CategoryProductsResponse {
  results: CategoryProductItem[];
  total: number;
  page: number;
  limit: number;
}

export interface HomeStats {
  total_products: number;
  pharmacy_count: number;
}

export interface PriceHistoryPoint {
  date: string;
  price?: number;
}

export interface PharmacyPriceHistory {
  pharmacy_source: string;
  data_points: PriceHistoryPoint[];
}

export interface PriceHistoryData {
  barcode: string;
  product_name?: string;
  history: PharmacyPriceHistory[];
}

File: backend/app/schemas/product.py
from datetime import datetime

from pydantic import BaseModel


class PharmacyPrice(BaseModel):
    pharmacy_source: str
    site_code: str | None = None
    product_name: str | None = None
    current_price: float | None = None
    original_price: float | None = None
    discount_percentage: float | None = None
    discount_amount: float | None = None
    bank_discount_price: float | None = None
    bank_discount_bank_name: str | None = None
    bank_payment_offers: str | None = None
    requires_prescription: bool | None = None
    product_url: str | None = None
    image_url: str | None = None
    scraped_at: datetime | None = None


class ComparisonResponse(BaseModel):
    barcode: str
    product_name: str | None = None
    brand: str | None = None
    main_category: str | None = None
    image_url: str | None = None
    best_price: float | None = None
    highest_price: float | None = None
    savings: float | None = None
    pharmacies: list[PharmacyPrice] = []


class SearchResultItem(BaseModel):
    group_key: str | None = None
    barcode: str | None = None
    product_name: str | None = None
    brand: str | None = None
    image_url: str | None = None
    main_category: str | None = None
    requires_prescription: bool | None = None
    best_price: float | None = None
    pharmacy_count: int | None = None


class SearchResponse(BaseModel):
    results: list[SearchResultItem] = []
    total: int = 0
    page: int = 1
    limit: int = 20


class CategoryItem(BaseModel):
    main_category: str
    product_count: int
    pharmacy_count: int | None = None
    min_price: float | None = None
    max_price: float | None = None


class PriceHistoryPoint(BaseModel):
    date: str
    price: float | None = None


class PharmacyPriceHistory(BaseModel):
    pharmacy_source: str
    data_points: list[PriceHistoryPoint] = []


class PriceHistoryResponse(BaseModel):
    barcode: str
    product_name: str | None = None
    history: list[PharmacyPriceHistory] = []
